README.md
autoBnk
=======

Скрипт для выборки сумм налогов, сборов, прочих обязательных платежей, контролируемых Министерством доходов и сборов, из файлов ГКС (приказ ГКУ/ГНСУ [№74/194 від 25.04.2002](http://zakon4.rada.gov.ua/laws/show/z0436-02))


CHANGES
29.04.14
платеж 1901 (65%) в госбюджете разбит на две части в общий и 
специальный фонды уходит соответственно 31,6% и 33,4%.
В связи с этим платеж будет обозначен в конфигурационных файлах как:

* 190110 - общий фонд
* 190120 - специальный фонд

26.08.14
добавлен военный сбор 11011000

27.10.14
добавлены радиочастоты

05.12.14
* добавлены Авансы по 
	1102 31 
	1102 33
	1102 40
	1102 46
	с общим кодом в 110231 в tax.xml

* добавлена возможность использовать курсив - параметр 
	/summary/divs/italic в summary.xml

10.02.15
* изменена процентовка
* добавлены:
  - в г/б -
  - НДФЛ 
  - НДФЛ в виде %

17.02.15
* добавлены строки «в т. ч.» по прибыли для коммунальных предприятий 
  а также для авансов.
* добавлена возможность создания БД на диске, для чего добавлен ключ 
  запуска memory [0|1]
* операции записи в БД теперь проводятся после всех транзакций, что 
  увеличивает быстродействие. 
* оптимизация алгоритма в части формирования дат при выводе.


dbftr.py
#!/usr/bin/env python 
# -*- coding: utf-8 -*-

# 'dbftr' -- DBF handle import class 
# 
# почему-то в этом файле дескриптор полей начинается с 32-го байта, и описание 
# поля в дескрипторе полей занимает тоже 32 байта.

class dbfToList:
	# позиции байтов для считывания соответственно количество записей, 
	# количество байт в заголовке, и количество байт в записи. 
	def __init__(self, dbfname):
		self.dbf = dbfname
		self.bytelist = (4,2,2)
		self._list = []
		self.dfile = open(self.dbf,'rb')
		self._readbytes()

	def read_table(self):
		return self.read_and_split()
		
	def _readbytes(self):
		"""переменная _list будет содержать:
		[<количество записей>,<число байт в заголовке файла>,<число байт поля>]
		В случае с казначейской таблицей это [69, 257, 56]
		"""
		# с 4й позиции начинается число записей
		self.dfile.seek(4)
		for i in self.bytelist:
			self._list.append(self.dfile.read(i))
		for i in range(len(self._list)):
			self._list[i]=int.from_bytes(self._list[i],byteorder='little')
		
	def _get_fields(self):
		self.dfile.seek(32)
		_fielddesc = []
		for i in range(int(self._list[1]-2-32)):
			if i % 32 == 0:
				_p = self.dfile.read(32)
				field_desc = _p
				_fielddesc.append(tuple([field_desc[:4].decode().split('\x00')[0],\
					field_desc[11:12].decode(),int.from_bytes(field_desc[16:17],\
					byteorder='little')]))
		return (_fielddesc)

	def read_and_split(self):
		# получаем имена и длины и типы полей
		# _th = self._get_fields()
		# список _rowlist сюда всё запишем
		rowlist = []
		self.dfile.seek(self._list[1])
		for i in range (self._list[0]):
			g = self.dfile.read(56).decode()[1:]
			# сильное колдунство -- нас изначально известны нужные поля
			# ['2', ['1905', '03', '00'], 0, 0]
			rowlist.append([g[11:12],[g[15:23][:4],g[15:23][4:6], \
				g[15:23][6:8]], int(g[23:40]), self.dbf.split('\\')[-1][2]])
		return rowlist
		

autobnk4.py
""" 
AutoBnk4 main 
version 4.0

  The MIT License (MIT)
  Copyright (c) 2013 Renat Nasridinov, <mavladi@gmail.com>
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE. """

import sqlite3
import xml.etree.ElementTree as ET
import os.path
from os import listdir
from os.path import isfile, join


"""
Глобальные списки, константы и прочее
"""
# константа госбюджет
DB = 0
# константа местный бюджет
MB = 1

tr_ext = []    # кортеж для расширений файлов казны
raj_dict = {}  # словарь сопоставления казначейств районам

class db_processing:
	def __init__(self):
		# база данных в памяти
		self.engine = sqlite3.connect(':memory:')
		print('>> Engine created.')
		self.db_cur = self.engine.cursor()

#"CREATE TABLE itog (code text, raj integer, zn integer)")

	def cross_process(self):
		#self.db_cur.execute("ALTER TABLE etalon RENAME TO temp_names")
		self.db_cur.execute("INSERT INTO itog \
								SELECT code AS code, raj AS raj, SUM(zn) as zn \
								FROM itog_tmp GROUP BY code, raj")
		self.db_cur.execute("CREATE TABLE bank_sum (code text, raj83 integer, raj87 integer, raj18 integer)")
		self.db_cur.execute("INSERT INTO bank_sum \
								SELECT u.code, \
										sum(s83.zn) as raj83, \
										sum(s87.zn) as raj87, \
										sum(s18.zn) as raj18 \
								FROM etalon u \
									left outer join \
										itog s83 on u.code = s83.code \
										and s83.raj = 83 \
									left outer join \
										itog s87 on u.code = s87.code \
										and s87.raj = 87 \
									left outer join \
										itog s18 on u.code = s18.code \
										and s18.raj = 18 \
								GROUP BY u.code")
		self.db_cur.execute("SELECT e.nompp, e.name, \
									s.raj83, s.raj87, s.raj18 \
									from etalon e \
									left outer join \
									bank_sum s \
									on e.code = s.code\
									ORDER BY nompp")
		return(self.db_cur.fetchall())

	def make_etalon(self):
		etalon_row = ET.parse('config\\etalon.xml')
		table = etalon_row.getroot()
		for row in table:
			#print('insert into etalon (code, name) values ("%s", "%s")' % (row[0].text, row[1].text))
			self.db_cur.execute('insert into etalon (code, name, nompp) values ("%s", "%s", %d)' % (row[0].text, row[1].text, int(row[2].text)))
			self.engine.commit()

	def create_tables(self, raj_list):
		# создание таблиц
		self.db_cur.execute("CREATE TABLE bank (raj integer, rozd text, rd text, pg text, st text, zn integer, bd integer)")
		self.db_cur.execute("CREATE TABLE itog_tmp (code text, raj integer, zn integer)")
		self.db_cur.execute("CREATE TABLE itog (code text, raj integer, zn integer)")
		self.db_cur.execute("CREATE TABLE etalon (code text, name text, nompp integer)")
		#self.db_cur.execute("CREATE TABLE out (code text, name text, nompp integer)")
		self.engine.commit()

	def fill_table(self, tr_values, raj_code):
		# заполнение таблицы
		for e in tr_values:
			self.db_cur.execute('insert into bank values (%s, %s, "%s", "%s", "%s", %s, %s)' % 
				(raj_dict[raj_code],e[0],e[1][0],e[1][1],e[1][2],e[2],str(e[3])))
			self.engine.commit()

	def list_tables(self):
		# перечисление таблиц
		self.db_cur.execute("SELECT name FROM sqlite_master WHERE type='table' ORDER BY name;")
		print(self.db_cur.fetchall())

	def retrieve_table(self, table_name):
		# вывод таблицы по имени
		print("Таблица: ", table_name)
		self.db_cur.execute("SELECT * FROM "+table_name)
		for element in self.db_cur.fetchall():
			print(element)
		print('-'*80)

	def sql_construct(self,code,params):
		rozd = ('' if params[0]==None else " AND rozd='"+params[0]+"'")
		bd = params[1] # всегда присутствует
		rd = params[2] # всегда присутствует
		pg = ('' if params[3]==None else " AND pg='"+params[3]+"'")
		coef = params[4] # всегда присутствует
		query = "SELECT raj, SUM(zn) * %f as 'zn' FROM bank WHERE bd=%s AND rd='%s' %s %s GROUP BY raj;" % (float(params[4]), params[1], params[2], pg, rozd)
		self.db_cur.execute(query)
		for e in self.db_cur.fetchall():
			# район, код, сумма
			self.db_cur.execute("insert into itog_tmp values (%s, %s, %d)" % (code, e[0], e[1]))


	def processing(self):
		tax_list = ET.parse('config\\tax.xml')
		tax = tax_list.getroot()
		for row in tax:
			for query in row:
				params = []
				# переменная c - от "компонент"
				# <rozd> <bd> <rd> <pg> <coef>
				for c in query:
					params.append(c if c == None else c.text) 
				self.sql_construct(row.attrib['code'],params)

def parse_file(tr_dir, tr_f):
	# получаем имя файла, соединяя папку и имя
	tr_file_name=os.path.join(tr_dir, tr_f)
	# определяем тип бюджета
	if tr_f[2]=='0':
		bd_type=DB
	else:
		bd_type=MB
	tr_val = []
	with open(tr_file_name, encoding='cp866') as tr_file:
		tr_lines = tr_file.readlines()
		tr_file.close()
	for i in range(len(tr_lines)):
		e = tr_lines[i].split()
		if len(e)>4 and e[1].isdigit():
			del(e[1])
			del(e[-1])
			e[1]=[e[1][0:4],e[1][4:6],e[1][6:]]
			e[-1]=float(e[-1].replace("'",""))*100
			e.append(bd_type)
			tr_val.append(e)
	base.fill_table(tr_val, tr_f[-3:])

def config():
	if os.path.exists('config\\config.xml'):
		treasury_conf = ET.parse('config\\config.xml')
		tr = treasury_conf.getroot()
		bank_dir=tr[0][0].text
		if not os.path.exists(tr[0][1].text):
			os.makedirs(tr[0][1].text)
		for tr_code in tr.iter('code'):
			tr_ext.append(tr_code.text)
		for item in tr.iter('file'):
			raj_dict[item[1].text]=item[2].text
		return bank_dir

def make(bankpath):
	base.create_tables(tr_ext)
	# Создание списков файлов 
	for i in [ f for f in listdir(bank_dir) if isfile(os.path.join(bank_dir,f)) ]:
		parse_file(bank_dir,i)

if __name__=="__main__":
	base = db_processing()
	bank_dir=config()
	make(bank_dir)
	base.processing()
	base.make_etalon()
	out = base.cross_process()
	print('out', out)
	#base.retrieve_table('bank')
	#base.retrieve_table('itog')
	#base.retrieve_table('bank_sum')
	#base.retrieve_table('etalon')

utils.py
#!/usr/bin/env python 
# -*- coding: utf-8 -*-

"""
autobnk_utils
Copyright (c) 2008 - 2015 Renat Nasridinov, <mavladi@gmail.com>

Модуль вспомогательных утилит для autobnk.py
Распространяется на тех же условиях, что и autobnk.py
"""

class dbfToList:
    """Класс считывания таблицы из файла DBF в список. 
    Возвращает список, в который вложены списки, являющиеся строками исходной 
    таблицы.
    Примечание. В файлах из ГКСУ дескриптор полей начинается с 32-го байта, и 
    описание поля в дескрипторе полей занимает тоже 32 байта (позиции байтов 
    для считывания соответственно количество записей, количество байт в 
    заголовке, и количество байт в записи. 
    """
    def __init__(self, dbfname):
        """Параметры:
            dbfname - имя файла DBF
        """
        self.dbf = dbfname
        self.bytelist = (4,2,2)
        self._list = []
        self.dfile = open(self.dbf,'rb')
        self._readbytes()

        
    def _readbytes(self):
        """переменная _list будет содержать:
        [<количество записей>,<число байт в заголовке файла>,<число байт поля>]
        В случае с казначейской таблицей это [69, 257, 56]
        """
        # с 4й позиции начинается число записей
        self.dfile.seek(4)
        for i in self.bytelist:
            self._list.append(self.dfile.read(i))
        for i in range(len(self._list)):
            self._list[i]=int.from_bytes(self._list[i],byteorder='little')
        
    def _get_fields(self):
        self.dfile.seek(32)
        _fielddesc = []
        for i in range(int(self._list[1]-2-32)):
            if i % 32 == 0:
                _p = self.dfile.read(32)
                field_desc = _p
                _fielddesc.append(tuple([field_desc[:4].decode().split('\x00')[0],\
                    field_desc[11:12].decode(),int.from_bytes(field_desc[16:17],\
                    byteorder='little')]))
        return (_fielddesc)


    def read_table(self):
        """Непосредственно парсит таблицу и возвращает переменную _rowlist, 
        которая является списком списков вида
        ['2', ['1905', '03', '00'], 0, 0]
        """
        # получаем имена и длины и типы полей:
        # _th = self._get_fields()
        rowlist = []
        self.dfile.seek(self._list[1])
        for i in range (self._list[0]):
            g = self.dfile.read(56).decode()[1:]
            # сильное колдунство -- y нас изначально известны нужные поля:
            rowlist.append([g[11:12],[g[15:23][:4],g[15:23][4:6], \
                g[15:23][6:8]], int(g[23:40]), self.dbf.split('\\')[-1][2]])
        return rowlist


